README.md
# Tarantool-http OpenAPI support

You may use your openapi specification file to map the routes and handlers inside your project.

## Table of contents

- [Quickstart](#quickstart)
- [Logging](#logging)
- [Routing](#routing)
- [Security Example](#security-example)
- [Describing multipart/formdata request](#describing-multipartform-data-request)
- [Route settings](#route-settings)
- [Check schema coverage](#check-schema-coverage)
- [Strict mode](#strict-mode)
- [CORS handling](#cors-handling)

## Quickstart

```lua
local openapi = require("gtn.openapi")

local app = openapi(
    require("http.server"),
    require("http.router"),
    "api.yaml",
    {
        security = require("authorization")
    }
)

app:start()
```

As you can see, openapi call takes four positional arguments, which are:
server module, router module, path to read specification from and also a table with
options.

## Logging

To log requests, errors and also show error messages on the client side, you could pass the debug options:

```lua
local app = openapi(
    require("http.server"),
    require("http.router"),
    "api.yaml",
    {
        security = require("authorization"),
        debug = true
    }
)
```

This option will enable the logging and write it to the sysout by default.
To pass a logfile or logging level, just add a "log" table to the options:

```lua
local app = openapi(
    require("http.server"),
    require("http.router"),
    "api.yaml",
    {
        security = require("authorization"),
        log = {
            file = "log/app.log",
            level = 6
        },
        debug = true
    }
)
```

## Routing

The openapi module uses `opeartionId` and `tags` options from [operation object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#operation-object)
to map controller function to the route. For example, you wrote a `./controllers/example.lua` module:

```lua
local _M = {}
function _M.userinfo(self)
    -- some code here
    return self:render({
        json = {
            data = user_data
        }
    })
end

return _M
```

Operation object from spec file should look something like this:

```yaml
/example:
  get:
    tags:
      - example
    operationId: "userinfo"
    summary: "Example request"
    # ...
```

So, the module will map `GET /example` request to be handled by `userinfo` function in `controllers.example` module.
All the modules must be stored within the `controllers` directory, because the module exploits tarantool-http's ability
to map handlers with a `controller#action` string. Place your controller modules inside the **controllers** folder, relative
to your `app_dir` option.

The controller module may also return a function instead of a table, in that case just make sure your `example.lua` module
returns the function and drop the operationId from operation object schema.

Note that the module maps only the first tag of the list, every other tag would be ignored.

## Security Example

The `security` option contains either a table of methods to handle security schemas described
in your spec-file, or a function if you don't have multiple authorization protocols.

api.yaml

```yaml
components:
securitySchemes:
  bearerAuth:
    type: http
    scheme: bearer
  apiKeyAuth:
    type: apiKey
    in: header
    name: X-API-KEY
  basicAuth:
    type: http
    scheme: basic
  cookieAuth:
    type: apiKey
    in: cookie
    name: cookiename
```

authorization.lua file

```lua
local _M = {}

function _M.bearerAuth(url, scopes, token)
    -- validate token
    return result, err
end

function _M.apiKeyAuth(url, scopes, api_key)
    -- validate API-key
    return result, err
end

function _M.basicAuth(url, scopes, username, password)
    -- validate user credentials
    return result, err
end

function _M.cookieAuth(url, scopes, cookie)
    -- validate cookie here
    return result, err
end

return _M
```

The module correlate those automatically by the name of security scheme and method.
and also sends corresponding arguments to the function. The basic authorization header is
decoded from base64 and also splits by ":" symbol to form username and password.

The `url` argument is the current request's path, taken from `request_object.req.path`.
The scopes are sent from the security option in the operation object schema:

```yaml
get:
  tags:
    - example
  operationId: "userinfo"
  summary: "Example request"
  security:
    - bearerAuth: ["test_scope", "example_scope", "etc_scope"] #here they are
```

As for token, api_key and username\password pair: those are, obviously, your authorization data,
that needs to be checked in order to proceed with the request handling process.

Authorization functions must return two values. In case of an error, be sure to `return nil, error_string`,
in case of successful authorization just return current user's data, that you may later access inside your
controller function. For example, inside of our `controllers/example.lua` described above:

```lua
function userinfo(self)
    local user_data = self.authorization

    return self:render({
        json = {
            data = user_data
        }
    })
end
```

You may also override default error handling with a function:

```lua
local openapi = require("gtn.openapi")

local app = openapi(
    require("http.server"),
    require("http.router"),
    "api.yaml",
    {
        security = require("authorization")
    }
)

-- the default error override
-- invalid repsponse format, or some crucial option that is not set for the endpoint will end up here
app:default(
    function(ctx, err)
        return ctx:render({
            status = 204,
            json = {
                success = false,
                message = err,
                error   = "No Content"
            }
        })
    end
)

-- all unexpected errors during the call of the actual handler will end up here
app:error_handler(
    function(ctx, err)
        -- err argument here will be a table most of the time
        return ctx:render({
            json = {
                success = false,
                errors  = err
            }
        })
    end
)

-- the second return value of our `bearerAuth`, `apiKeyAuth` and `basicAuth` functions will be here
app:security_error_handler(
    function(ctx, err)
        return ctx:render({
            status = 401,
            json = {
                success = false,
                error   = "Unauthorized",
                message = err
            }
        })
    end
)

-- override the default 404 handler if needed
-- 404.html file form the "/templates" folder will be rendered by default
-- the first parameter is either a function, or a boolean value
-- the second parameter is the matching pattern
app:not_found_handler(
    function(ctx)
        return ctx:render({
            json = {
                success = false,
                error   = "Not found"
            }
        })
    end,
    "/api/v1/*path"
)

-- to render default 404 for any unmatched path
app:not_found_handler(true)


-- all of the request parameters validation errors would end up here
-- by default the response in json = { error = err } with http-status of 400
app:bad_request_handler(
    function(ctx, err)
        ctx:render({
            status = 400,
            json = {
                success = false,
                error   = err,
                msg     = "Bad request"
            }
        })
    end
)
```

## Describing multipart/form-data request

OpenAPI specification has two options to describe a file parameter for now:

```yaml
# this one is for raw binary file data
/upload/binary:
  post:
    tags:
      - main
    operationId: "upload_binary"
    requestBody:
      content:
        multipart/form-data:
          schema:
            type: object
            properties:
              file:
                type: string
                format: binary

# this one is for base64-encoded binary
/upload/bytes:
  post:
    tags:
      - main
    operationId: "upload_bytes"
    requestBody:
      content:
        multipart/form-data:
          schema:
            type: object
            properties:
              file:
                type: string
                format: bytes
```

Take note, that inside of an openapi validator, the _bytes_ and _binary_ formats are actually treated
as an object and not a string value. So, inside of a controller function the **file** parameter will be a table containing:

- **data** — binary file content
- **headers** — file headers, containing _filename_ and _name_ which are: actual file name(~duh) and parameter name accordingly
- **mime** — a mime type of the file: image/jpeg, image/png etc.

## API versioning

There's a possibility to pass multiple schemas to the openapi object constructor function. It's primarily aimed to add
an ability to maintain several versions of your API.

```lua
-- app.lua file
local openapi = require("gtn.openapi")

local schema_options = {
    base_path         = "schemas",
    primary_schema    = "base.yaml",
    secondary_schemas = {
        {
            schema = "api_v2.yaml",
            path = "/api/v2",
        },
        {
            schema = "api_v3.yaml"
        },
        {
            schema = "relative.yaml",
            path   = "/relative",
            relative = true
        }
    }
}

local app = openapi(
    require("http.server"),
    require("http.router"),
    schema_options,
    {
        security = require("authorization")
    }
)

app:start()
```

Let's take a closer look to _schema_options_ table fields:

- **base_path** — option indicates in which directory the schemas will be located, in this case the server will look for
  a schemas folder relatively to your _app_dir_ option
- **primary_schema** — quite self-descriptive: the primary schema from which the server settings and all of the basic
  paths and components will be taken
- **secondary_schemas** — a list of additional schemas, in this case a v2 and v3 of our API.
- **schema** — in secondary_schemas is a name of a file relative to our **base_path** option, i.e **./schemas/v2.yaml** and **./schemas/v3.yaml** in this case.
- **path** — is a prefix to all the paths, described in current schema. If this option is not set, this file will simply extend the primary schema.
  Please note, that if there's a global path option set in primary schema, it'll also be applied to this one's paths.
- **relative** - is and option that idicates, that given schema paths would be relative to primary ones, i.e. the primary
  schema's base_path would be a prefix to every path withing this schema

You may also set new schema without changing the old-way options, just by calling a couple of new methods:

```lua
-- app.lua file
local app = openapi(
    require("http.server"),
    require("http.router"),
    "schema.yaml",
    {
        security = require("authorization")
    }
)

--[[
   the first argument is the path to the schema file to read
   the second one is actually a base_path option from before.
   this calls will add new secondary schemas inside of openapi object
]]
app.openapi:add_schema("./schemas/api_v2.yaml", "/api/v2")
app.openapi:add_schema("./schemas/api_v3.yaml")

-- calling this method will automatically set new routes to server object
app:bind_paths()
```

## Route settings

There are additional tweaks for a single path object, that are set in a special option, called **x-settings**. For example,
you have a global prefix for your API endpoints:

```yaml
servers:
  - url: http://localhost:{port}/{path}/
  description: 'development'
  variables:
    port:
      default: '8080'
    path:
      default: 'api/v1' # this one here
```

And you need just a couple of enpoints to ignore this options and be outside of "/api/v1' path. It's simple: just add
an x-settings option with fullPath inside of those:

```yaml
/outside_path:
  get:
    x-settings:
      fullPath: true
    responses:
      200:
        desciption: I am outside
/outside_too:
  post:
    x-settings:
      fullPath: true
    responses: ...
```

It primarily has some features for automatic schema testing, that are initiated by **./app.lua --test** command. For, example,
if you want to exclude some parts of api from being tested or you expect some other http-code in response:

```yaml
  /outside_path:
    get:
      x-settings:
        fullPath: true
        skipTest: true # this endpoint will skip testing
        strict: true
  /outside_too:
    post:
      x-settings:
        fullPath: true
        testStatus: 403 # Will look for 403 schema in responses part
    responses:
      # not this one
      200:
        desciption: Ok
      # yup, this one
      403:
        description: Bad Request
          content:
            application/json:
              schema:
                type: object
                required:
                  - success
                properties:
                  success:
                    type: boolean
                    # will expect this value in assertion
                    example: true
                  error:
                    type: string
                    # this field is optional, but if exists, this value will be expected during auto testing
                    example: invalid parameters
```

The **strict** option determines if it should check the response value for equality with an example value
or should it just check that the response types match with those set in the schema.

## Check schema coverage

There's a command to check schema controller coverage, i.e. to determine which parts of the schema do not have actual
controllers written. For example, you wrote a simple schema with 3 paths:

```yaml
paths:
  /this_one:
    get:
      tags:
        - example
      operationId: this_one
      responses:
        200:
          description: ok
  /that_one:
    get:
      tags:
        - example
      operationId: that_one
      responses:
        200:
          description: ok
  /another_one:
    get:
      tags:
        - example
      operationId: another_one
      responses:
        200:
          description: ok
```

Also, you have **example.lua** file inside your project's **controllers** directory:

```lua
-- example.lua
local _M = {}

function _M.this_one(self)
    return self:render({
        json = {
            success = true
        }
    })
end

return _M
```

And you want to find parts of your API not covered by handlers. Just run the command:

```bash
./app.lua --test coverage
FAILED TOTAL: 2
FAILED PATHS:
/that_one
/another_one
```

See, there are two endpoints described in openapi schema, but not actually handled. This feature is particularly handy if you have a
massive several-thousand-lined schema.

## Strict mode

To set the strict parameter validation mode set the strict parameter in options to true:

```lua
local app = openapi(
    require("http.server"),
    require("http.router"),
    "schema.yaml",
    {
        security = require("authorization"),
        strict   = true
    }
)
```

## CORS handling

You may set cors handling by setting the **cors** option to your openapi instance initialization

#### Example

```lua
local openapi = require("gtn.openapi")

local app = openapi(
    require("http.server"),
    require("http.router"),
    "api.yaml",
    {
        security = require("authorization"),
        cors     = {
            -- default value 3600
            max_age = 18400,
            -- default value true
            allow_credentials = false,
            -- default {"Authorization", "Content-Type"}
            allow_headers = {"Authorization", "Content-Type", "X-Requested-With"},
            -- default {"*"}
            allow_origin = {"http://example.com"}
        }
    }
)

app:route({path="/api/user", method="GET"}, some_handler)

app:start()
```

To set default CORS settings:

```lua
local openapi = require("gtn.openapi")

local app = openapi(
    require("http.server"),
    require("http.router"),
    "api.yaml",
    {
        security = require("authorization"),
        cors     = {} -- takes only a table value
    }
)
app:start()
```
The default CORS options should suffice for the development process.

There's also a way to override CORS handler for specific endpoints:
```lua
local app = openapi(
    require("http.server"),
    require("http.router"),
    "api.yaml",
    {
        security = require("authorization"),
        cors = {
            max_age = 18600,
            allow_credentials = false,
            allow_headers = {"Authorization", "Content-Type", "X-Requested-With"},
            allow_origin = {"http://example.com"},
            specific = {
                {
                    -- this option obviously determines which path the current settings will be applied to
                    path = "/some_path",
                    -- this option disables the method check for current endpoint
                    -- i.e it won't intercept the control if requested method is not allowed
                    -- and also it won't set the self.hdrs table automatically
                    skip_method = false,
                    -- a function to override the cors handling
                    -- you may leave it empty if there's a need to skip CORS handling at all
                    handler = function(self)
                        self.hdrs["access-control-allow-origin"] = "example.com"
                        self.hdrs["access-control-allow-headers"] = "Authorization,Content-Type"

                        return self:render({
                            status = 201,
                            text = ""
                        })
                    end
                }
            }
        }
    }
)
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## \[2.0-1\] - 2021-08-02

### Fixed
- a multipart handler bug

## \[2.0-0\] - 2021-06-28

### Fixed
- project module structure
- error metrics

## \[1.1-8\] - 2021-06-21

### Fixed
- a very serious bug regarding error interception, that led to stack overflow

## \[1.1-7\] - 2021-05-31

### Added
- x-settings **strict** option to just check the type of the response values on schema auto-testing

## \[1.1-6\] - 2021-04-05 13:46

### Fixed
- migrated deprecated prometheus lib to the metrics lib
- not valid request body exception output

## \[1.1-5\] - 2021-03-25 12:00

### Added 
- Specific endpoints CORS options override

## \[1.1-4\] - 2021-01-29 11:04

### Added

- debug option
- log option
- logging description to the readme file

## \[1.1-3\] - 2020-10-16

### Added

- strict mode for schema paramter invalidation
- changelog


run_tests.sh
#!/usr/bin/env sh

for file in ./tests/*
do
  tarantool $file
done

tests/test_util.lua
-- tnt modules
local tap  = require("tap")

local util = require("src.util")

local testing = tap.test("util module")

testing:plan(4)

local parsed_path = "http://example.com/:test/:test1/:test2"

local mock_header_function = function(self, name)
    return self[name]
end

local mock_bearer_ctx = {
    authorization = "Bearer test",
    header = mock_header_function
}

local mock_basic_ctx = {
    authorization = "Basic dGVzdDp0ZXN0",
    header = mock_header_function
}

local mock_api_key_ctx = {
    ["API-KEY"] = "test",
    cookie = "SomeValue=Wrong; Test=Test",
    header = mock_header_function
}

local mock_cached_query_param = {
    query_params = {
        test = "test"
    }
}

testing:is_deeply(
    util.read_path_parameters("http://example.com/{test}/{test1}/{test2}"),
    {"test", "test1", "test2"},
    "read_path_parameters true"
)

testing:is(
    util.parse_path("http://example.com/{test}/{test1}/{test2}"),
    parsed_path,
    "parse_path true"
)

testing:is(
    util.bearer(mock_bearer_ctx),
    "test",
    "bearer true"
)

local basic_username, basic_password =  util.basic(mock_basic_ctx)
testing:is(
    basic_username,
    "test",
    "basic username"
)
testing:is(
    basic_password,
    "test",
    "basic password"
)

testing:is(
    util.apiKey(mock_api_key_ctx, "Test", "cookie"),
    "Test",
    "cookie true"
)

testing:is_deeply(
    util.cached_query_param(mock_cached_query_param, "test"),
    "test",
    "cached_query_param named"
)

testing:is_deeply(
    util.cached_query_param(mock_cached_query_param),
    {test = "test"},
   "cached_query_param unnamed"
)

tests/test_validation.lua
-- tnt modules
local tap        = require("tap")

local validation = require("src.validation")

local testing = tap.test("vaidation module")

testing:plan(21)

local enum_test_value = {"test", "test1", "test2"}
local p_error_format = {
    details = {
        error    = "test",
        expected = "test1",
        actual   = "test2"
    }
}

testing:ok(validation.is_object({}) == false, "is_object empty array")
testing:ok(validation.is_object({1}) == false,"is_object array")
testing:ok(validation.is_object({k = 1}) == true, "is_object object")

testing:ok(validation.string(false) == false, "string false")
testing:ok(validation.string("ok") == true, "string true")

testing:ok(validation.integer(false) == false, "integer false")
testing:ok(validation.integer(1) == true, "integer true")

testing:ok(validation.email("testing") == false, "email false")
testing:ok(validation.email("example@example.com") == true, "email true")

testing:ok(validation.uuid("abcdef") == false, "uuid false")
testing:ok(validation.uuid("66d9ca63-551e-4538-b59f-9f0f29522d70") == true, "uuid true")

testing:ok(validation.enum("test3", enum_test_value) == false, "enum false")
testing:ok(validation.enum( "test", enum_test_value) == true, "enum true")
testing:ok(validation.enum("test", "olegue") == false, "enum wrong type")

testing:ok(validation.date("123-123-123") == false, "date false")
testing:ok(validation.date("2021-07-05") == true, "date true")

testing:ok(validation.boolean(1) == false, "boolean false")
testing:ok(validation.boolean(true) == true, "boolean true")

testing:ok(validation['date-time']("123-123-123 20:20:20") == false, "date-time false")
testing:ok(validation['date-time']("2021-07-05 16:15:22") == true, "date-time true")

testing:is_deeply(
    validation.p_error("test", "test1", "test2"),
    p_error_format,
    "p_error format true"
)

src/util.lua
-- tnt modules
local base64_decode = require("digest").base64_decode
local http_utils    = require('http.utils')
local tsgi          = require("http.tsgi")
local fio           = require("fio")
local fun           = require("fun")
local lib           = require("http.lib")

local _M = {}

function _M.read_config()
    local env = arg[1] or "development"

    if env:startswith("--") then
        env = "development"
    end

    local status, conf = pcall(require, "config."..env)

    assert(
        status,
        ("Invalid environment %q. Try adding it to the config directory"):format(env)
    )
    conf.__name  = env

    _G.app_config = conf
    _G.app_config.is_test = fun.any(
        function(val)
            return val == "--test"
        end,
        arg
    )
end

function _M.read_path_parameters(path_str)
    local result = {}
    for val in path_str:gmatch("{(%w+)}") do
        table.insert(result, val)
    end
    return result
end

function _M.parse_path(p)
    local res = p
    for path_param in p:gmatch("{(%w+)}") do
        local sub = ("{%s}"):format(path_param)
        res = res:gsub(sub, ":"..path_param)
    end
    return res
end

-- basic responses
function _M.not_implemented(self, tag, operation_id)
    return self:render({
        status = 501,
        json = {
            error       = "Not Implemented",
            tag         = tag,
            operationId = operation_id
        }
    })
end

function _M.read_specfile(filepath)
    assert(
        fio.path.is_file(filepath),
        ("Spec file %s does not exist"):format(filepath)
    )

    local tab = filepath:split(".")
    local ext = tab[#tab]
    assert(ext == "yaml" or ext == "json", "Invalid openapi file extension")
    local decoder = require(ext)
    local file = fio.open(filepath, {"O_RDONLY"})

    local data = file:read()
    local status, res = pcall(decoder.decode, data)
    assert(
        status,
        ("Specification file %s is not valid"):format(filepath)
    )

    return res
end

--[[
    initial primary schema mutation. NOTE: this is not copying
    it changes the original schema
]]
function _M.join_schemas(primary, secondary, field)
    primary[field]   = primary[field] or {}
    secondary[field] = secondary[field] or {}

    primary[field] = fun.reduce(
        function(res, k ,v)
            -- does not override primary options with secondary ones
            if not res[k] then
                rawset(res, k ,v)
            end

            return res
        end,
        primary[field],
        secondary[field]
    )
end

function _M.bearer(ctx)
    local header = ctx:header("authorization")

    if not header then
        return
    end

    return header:match("Bearer (.*)")
end

function _M.basic(ctx)
    local header = ctx:header("authorization")

    if not header then
        return
    end

    local creds = base64_decode(header:match("Basic (.*)"))
    if not creds then
        return
    end
    return creds:match("(%w+):(%w+)")
end

function _M.apiKey(ctx, name, goes_in)
    if goes_in == "header" then
        local n = name:lower()
        return ctx:header(n)
    elseif goes_in == "cookie" then
        local val = ctx:header("cookie")

        return val and val:match(("%s=([^;]*)"):format(name)) or nil
    end
end

-- bad unescape fpr query parameters
function _M.cached_query_param(self, name)
    if name == nil then
        return self.query_params
    end
    return self.query_params[ name ]
end

function _M.query_param(self, name)
    if self:query() ~= nil and string.len(self:query()) == 0 then
        rawset(self, 'query_params', {})
    else
        local params = lib.params(self['QUERY_STRING'])
        local pres = {}
        for k, v in pairs(params) do
            pres[ http_utils.uri_unescape(k) ] = http_utils.uri_unescape(v, true)
        end
        rawset(self, 'query_params', pres)
    end

    rawset(self, 'query_param', _M.cached_query_param)
    return self:query_param(name)
end

-- overrides response render handling
function _M.render(self)
    local render_func = self.render
    self.render = function(ctx, input)
        local resp = render_func(ctx, input)

        resp.status = input.status

        if ctx.hdrs then
            for k, v in next, ctx.hdrs do
                resp.headers[k] = v
            end
        end

        if input.headers then
            for k, v in next, input.headers do
                resp.headers[k] = v
            end
        end

        return resp
    end
    self.render_swap = true
end

local ni_patterns = {
    [[Can't load module '(.*)': '(.*)']],
    [[Controller '(.*)' doesn't contain function '(.*)']],
    [[require '(.*)' didn't return table]],
    [[Controller '(.*)' is not a function]]
}

-- overrides route handler with this one
function _M.handler(self)
    local handler_func = self.endpoint.handler

    self.endpoint.handler_swap = true
    self.endpoint.handler = function(ctx)
        local status, resp = pcall(handler_func, ctx)

        local httpd = ctx['tarantool.http.httpd']

        if not status then
            local tag, op_id
            for _, val in next, ni_patterns do
                tag, op_id = resp:match(val)

                if tag ~= nil then
                    break
                end
            end

            if tag or op_id then
                return _M.not_implemented(ctx, tag, op_id)
            end

            return httpd.error_handler(ctx, resp)
        end
        if not resp then
            return httpd.default(ctx, "No Content")
        end

        return resp
    end
end

local function request_multipart(self)
    local body = self:read_cached()

    local sep = "--"..self:header("content-type"):match("boundary=(.+)"):gsub("-", "%%-")
    local _, e = body:find(sep.."\r\n")
    local eor = false

    local t = {}

    while not eor do
        local _s, _e = body:find(sep, e)
        if body:endswith("--", _e, _e + 2) then
            eor = true
        end

        local param_part = body:match("%aontent%-%aisposition:.-; (.-)\r\n\r\n", e - 1)

        -- eh. simple fix
        if not param_part then
            param_part = body:match("CONTENT%-DISPOSITION:.-; (.-)\r\n\r\n", e - 1)
        end

        param_part = param_part:gsub(";", "")

        local mime_type = body:match("%aontent%-%aype: (.-)\r\n", e - 1)

        if not mime_type then
            mime_type = body:match("CONTENT%-TYPE: (.-)\r\n", e - 1)
        end

        local content = {}

        for key, val in param_part:gmatch("(.-)=\"(.-)\"") do
            rawset(content, key:strip(), val:strip())
        end

        local value = body:sub(e, _s - 1):match("\r\n\r\n(.-)\r\n$")

        rawset(
            t,
            content.name,
            content.filename and {data = value, headers = content, mime = mime_type} or value
        )

        _, e = _s, _e + 2
    end

    return t
end

-- solely to parse multipart
function _M.post_param(ctx, name)
    local params = request_multipart(ctx)

    return name and params[name] or params
end

function _M.handle_cors(ctx)
    local httpd = ctx['tarantool.http.httpd']
    local router = ctx:router()

    if not ctx.render_swap then
        _M.render(ctx)
    end

    ctx.hdrs = {}

    local req_method = ctx:header("access-control-request-method") or ctx:method()

    local route = router:match(req_method, ctx:path())

    if not route then
        if ctx:method() == "OPTIONS" then
            return ctx:render({
                status = 201,
                text = ""
            })
        end
        return httpd.bad_request_handler(ctx)
    end

    ctx.endpoint = route.endpoint
    ctx.stash    = route.stash

    if not ctx.endpoint.handler_swap then
        _M.handler(ctx)
    end

    if fun.any(function(v) return v=="*" end, httpd.options.cors.allow_origin) then
        ctx.hdrs["access-control-allow-origin"] = "*"
    else
        if fun.any(function(v) return v==ctx:header("origin") end, httpd.options.cors.allow_origin) then
            ctx.hdrs["access-control-allow-origin"] = ctx:header("origin")
        end
    end

    ctx.hdrs['access-control-max-age'] = httpd.options.cors.max_age
    ctx.hdrs['access-control-allow-credentials'] = tostring(httpd.options.cors.allow_credentials)
    ctx.hdrs['access-control-allow-headers'] = table.concat(httpd.options.cors.allow_headers, ",")

    if ctx:method() == 'OPTIONS' then
        local methods
        if httpd.openapi then
            local path = httpd.openapi.paths[ctx.endpoint.openapi_path]

            if not path then
                return
            end

            methods = fun.map(
                function(k)
                    return k:upper()
                end,
                path
            ):totable()
        else
            methods = {ctx.endpoint.method}
        end

        if fun.any(function(val) return val == req_method end, methods) then
            ctx.hdrs['access-control-allow-methods'] = req_method
        end

        return ctx:render({
            status = 201,
            text = ""
        })
    end

    return tsgi.next(ctx)
end

function _M.bind_security(ctx)
    local self = ctx:router()

    local r = self:match(ctx:method(), ctx:path())
    ctx.endpoint = r.endpoint
    ctx.stash    = r.stash

    local httpd = ctx['tarantool.http.httpd']

    local security = ctx.endpoint.security or (ctx.endpoint.openapi_path and httpd.openapi.global_security)

    if security ~= nil and next(security) then
        local auth_handler
        if not httpd.options.security then
            return httpd.default(ctx, "Security options are not specified for this server instance")
        end

        if type(httpd.options.security) == "table" then
            auth_handler = httpd.options.security[security.name]
        else
            auth_handler = httpd.options.security
        end

        if auth_handler then
            if not security.options and ctx.endpoint.uid_schema then
                local _s  = httpd.openapi:get_secondary(ctx.endpoint.uid_schema)
                local _ps = assert(_s.paths[ctx.endpoint.openapi_path], "path unknown")
                local _ms = assert(_ps[ctx.endpoint.method:lower()], "unknown method")

                ctx.endpoint.security = httpd.openapi:parse_security_schemes(_ms.security)
                security = ctx.endpoint.security
            end

            local scheme = security.options.scheme or security.options.type
            local auth_data, additional = _M[scheme](ctx, security.options.name, security.options['in'])
            if not auth_data then
                return httpd.security_error_handler(ctx, "Authorization data not found")
            else
                local err
                ctx.authorization, err = auth_handler(ctx:path(),  security.scope, auth_data, additional)

                if err then
                    return httpd.security_error_handler(ctx, err)
                end
            end
        else
            return httpd.security_error_handler(ctx, "Security handler is not implemented")
        end
    end

    return tsgi.next(ctx)
end

function _M.httpd_default_handler(self, f)
    if type(f) ~= "function" then
        return self:render({
            status = 204,
            json = {
                error = f
            }
        })
    end

    self.default = function(ctx, err)
        return f(ctx, err)
    end
end

function _M.httpd_error_handler(self, f)
    if type(f) ~= "function" then
        error(f)
    end

    self.error_handler = function(ctx, err)
        local httpd = ctx['tarantool.http.httpd']
        if httpd.error_metrics and next(httpd.error_metrics) then
            httpd.error_metrics.errors:inc(1)
        end

        return f(ctx, err)
    end
end

function _M.httpd_security_error_handler(self, f)
    if type(f) ~= "function" then
        error(f)
    end

    self.security_error_handler = function(ctx, err)
        local httpd = ctx['tarantool.http.httpd']
        if httpd.error_metrics and next(httpd.error_metrics) then
            httpd.error_metrics.security_errors:inc(1)
        end

        return f(ctx, err)
    end
end

function _M.httpd_bad_request_handler(self, f)
    if type(f) ~= "function" then
        return self:render({
            status = 400,
            json = {
                error = f
            }
        })
    end

    self.bad_request_handler = function(ctx, err)
        return f(ctx, err)
    end
end

function _M.httpd_not_found_handler(ctx, f, match_pattern)
    local router = ctx:router()

    if ctx.http_404_swap then
        return
    end

    if f then
        local handler = function(self)
            local resp = type(f) == "function" and f(self) or self:render()
            resp.status = 404
            return resp
        end

        ctx.http_404_swap = true

        router:route(
            {
                method = "ANY",
                file   = "404.html",
                path   = match_pattern or "/*path"
            },
            handler
        )
    end
end

return _M
